### PR TITLE
ImmutableDB: check checksum when reading blocks

### PR DESCRIPTION
**Pull request by mrBliss**

### CODE REVIEW

**Comment by mrBliss** [comment]
Note that if this is merged, #1177 cannot yet be closed because of:

> Blocks in the volatile DB have not been checked for validity, but we have checked that the body hash in the header actually matches the body, so if that check fails later, it must point to a disk failure and we should trigger recovery.

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl.hs [hunk 1]
```diff
@@ -518,8 +522,8 @@ getEpochSlot hasFS hashInfo err curEpochInfo getWhat epochSlot =
           GetHeader -> fmap (headerHash,) $
               -- Get just the header
               withFile hasFS epochFile ReadMode $ \eHnd ->
-                -- TODO we cannot check the checksum in this case, as we're
-                -- not reading the whole block
+                -- We cannot check the checksum in this case, as we're not
+                -- reading the whole block
                 hGetExactlyAt hasFS eHnd size offset
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl.hs]: Introduce `hGetExactlyAtCRC` (?). That will do basically the same thing as we do now: read everything first and then compute the checksum. So we can later optimise this (or change it to allow incremental parsing) in one place.

File: ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl.hs [hunk 2]
```diff
@@ -518,8 +522,8 @@ getEpochSlot hasFS hashInfo err curEpochInfo getWhat epochSlot =
           GetHeader -> fmap (headerHash,) $
               -- Get just the header
               withFile hasFS epochFile ReadMode $ \eHnd ->
-                -- TODO we cannot check the checksum in this case, as we're
-                -- not reading the whole block
+                -- We cannot check the checksum in this case, as we're not
+                -- reading the whole block
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl.hs]: Should we make a more prominent reminder of this for ourselves for the future? If there is an error in the header, the first anyone would know would be the remote peer we serve the header to disconnecting from us for giving them a header with an invalid signature. So they would never request the full block and we'd not do the full crc check.

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl.hs]: Will do. I'm still planning to better document the new index format.

We could store a checksum of the header too, but that requires a new index format and migration.

**Pull request merged**